src/websocket.c

```diff
@@ -302,63 +403,88 @@ ws_parse_data(const char *frame, size_t sz, struct ws_msg **msg) {
 		p = frame + 10 + (has_mask ? 4 : 0);
 		if(has_mask) memcpy(&mask, frame + 10, sizeof(mask));
 	} else {
+		free(frame);
 		return WS_ERROR;
 	}
 
 	/* we now have the (possibly masked) data starting in p, and its length.  */
 	if(len > sz - (p - frame)) { /* not enough data */
+		evbuffer_prepend(ws->rbuf, frame, sz); /* put the whole frame back */
+		free(frame);
 		return WS_READING;
 	}
 
-	if(!*msg)
-		*msg = ws_msg_new();
-	ws_msg_add(*msg, p, len, has_mask ? mask : NULL);
-	(*msg)->total_sz += len + (p - frame);
+	if(out_msg) { /* we're extracting the message */
+		struct ws_msg *msg = ws_msg_new(frame_type);
+		ws_msg_add(msg, p, len, has_mask ? mask : NULL);
+		size_t processed_sz = len + (p - frame); /* length of data + header bytes between frame start and payload */
+		msg->total_sz += processed_sz;
+		*out_msg = msg;
+
+		evbuffer_prepend(ws->rbuf, frame + len, sz - processed_sz); /* remove processed data */
+	} else { /* we're just peeking */
+		evbuffer_prepend(ws->rbuf, frame, sz); /* put the whole frame back */
+	}
+	free(frame);
 
-	if(frame[0] & 0x80) { /* FIN bit set */
+	if(fin_bit_set) { /* FIN bit set */
 		return WS_MSG_COMPLETE;
 	} else {
 		return WS_READING;	/* need more data */
 	}
 }
 
-
 /**
  * Process some data just received on the socket.
+ * Returns the number of messages processed in out_processed, if non-NULL.
  */
 enum ws_state
-ws_add_data(struct http_client *c) {
+ws_process_read_data(struct ws_client *ws, unsigned int *out_processed) {
 
 	enum ws_state state;
+	if(out_processed) *out_processed = 0;
 
-	state = ws_parse_data(c->buffer, c->sz, &c->frame);
+	state = ws_peek_data(ws, NULL); /* check for complete message */
 
 	while(state == WS_MSG_COMPLETE) {
-		int ret = ws_execute(c, c->frame->payload, c->frame->payload_sz);
-
-		/* remove frame from client buffer */
-		http_client_remove_data(c, c->frame->total_sz);
+		int ret = 0;
+		struct ws_msg *msg;
+		ws_peek_data(ws, &msg); /* extract message */
+
+		if(msg->type == WS_TEXT_FRAME || msg->type == WS_BINARY_FRAME) {
+			ret = ws_execute(ws, msg);
+			if(out_processed) (*out_processed)++;
+		} else if(msg->type == WS_PING) { /* respond to ping */
+			ws_frame_and_send_response(ws, WS_PONG, msg->payload, msg->payload_sz);
+		} else if(msg->type == WS_CONNECTION_CLOSE) { /* respond to close frame */
+			ws->close_after_events = 1;
+			ws_frame_and_send_response(ws, WS_CONNECTION_CLOSE, msg->payload, msg->payload_sz);
+		} else {
+			char format[] =  "Received unexpected WS frame type: 0x%x";
+			char error[(sizeof format)];
+			snprintf(error, sizeof(error), format, msg->type);
+			slog(ws->http_client->s, WEBDIS_WARNING, error, 0);
```

Comment:
it would be good to have a test for this, although it would likely require it to be socket-based; I assume most websocket client libraries don't let users send invalid frames.